### DISCUSSION TITLE
Short circuit when getting optional values

### DISCUSSION THREAD

**Discussion by glcraft** [nushell/nushell#8502] [Polls]
I want to talk about the recent change for the new syntax to get optional values (https://github.com/nushell/nushell/pull/8379)

There is one line in examples that confuses me : 
```
{ foo: 123 }.bar?.baz # errors because `baz` is not present on the result from `bar?`
```
I think it should instead short circuit baz as bar is not required.

Let me explain :

In case I would do `$var.bar?.baz`, what I expect is bar is optional, but baz is required **if bar is present**. 
In other words : "if, in $var, there is no bar, return null. Otherwise, in bar, if there is no baz, this time, i want to raise an error"
Here is an example :
```
{ foo: 123 }.bar?.baz            # no bar, bar is optional so returns null
{ bar: 123 }.bar?.baz            # bar exists but no baz in bar. As baz is required, it raise an error
{ bar: { baz: 123 } }.bar?.baz   # baz present in bar, returns 123
{ foo: 123 }.bar.baz?            # no bar, bar is required, raise an error
{ bar: 123 }.bar?.baz?           # bar exists but no baz in bar. As baz is optional, returns null
```
With the actual syntax, we are required to write `$var.bar?.baz?` to get a null if bar is not present. But no matter how, the current line always returns null, it doesn't raise any error.

My syntax I described above give much more flexibility on getting optional or required value

What do you think ?

**Comment by Dorumin** [comment]
For what it's worth: other languages I know that have null access handling do short circuit

C#:
```cs
record Outer(Inner inner);
record Inner(string message);

Outer value = null;
outer?.inner.message; // null
```
JS:
```js
null?.inner.message; // undefined
```
Rust:
```rs
struct Outer { inner: Option<Inner> }
struct Inner { message: Option<String> }

fn example(outer: Option<Outer>) -> Option<String> {
    let message = outer?.inner.unwrap().message.unwrap();

    Some(message)
}

example(None)
```

Not to sway one way or the other, just in case familiarity is an important part of using this feature. Personally I like annotating each segment that can be null

**Comment by glcraft** [comment]
About the current implementation, with 
```
{ foo: 123 }.bar?.baz
```
If missing baz raise an error, no matter what bar is, why should we precise that bar is optional. In other words, with the current implementation, those 2 expressions is similar : 
```
{ foo: 123 }.bar?.baz
{ foo: 123 }.bar.baz
```

**Comment by rgwood** [comment]
I don't have a strong feeling either way. Happy to change the behaviour to short-circuit if people feel that that would be better.

**Comment by rgwood** [comment]
PR created: https://github.com/nushell/nushell/pull/8554
